### PR TITLE
Add a new API version apps/v1beta2

### PR DESCRIPTION
**Pull request by janetkuo**
xref: #49135
This PR adds a new API version `apps/v1beta2` which contains a copy (of types, conversions, and defaults) of `apps/v1beta1` StatefulSet, Deployment, and their subresources. Note that `apps/v1beta2` is still WIP and we will make breaking changes to it before releasing 1.8.

Moving core controllers (StatefulSet, Deployment, ReplicaSet, DaemonSet) to `apps/v1beta2` is the first step of moving them to `apps/v1` (GA). 

This PR is a starting point for DaemonSet and ReplicaSet to move from `/extensions` to `/apps` and for Deployment and StatefulSet to make some breaking changes (e.g. new defaults and/or remove deprecated fields).

```release-note
Add a new API version apps/v1beta2
```

### CODE REVIEW

**Comment by janetkuo** [comment]
@kubernetes/sig-apps-api-reviews 

This is ready for review.

File: pkg/apis/apps/v1beta2/conversion.go [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+limitations under the License.
+*/
+
+package v1beta2
+
+import (
+	"fmt"
+
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+
```
Comment (kargakis) [inline pkg/apis/apps/v1beta2/conversion.go]: kill this newline

**Comment by janetkuo** [comment]
/assign @smarterclayton 
need your approval

File: pkg/apis/apps/v1beta2/conversion.go (hunk 1 above)
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: Done

**Comment by janetkuo** [comment]
Rebased

**Comment by bgrant0607** [comment]
@janetkuo @kow3ns Where is the plan documented? Please link to it in the PR description.

This API is going to be changed, so it isn't ready to be used. Therefore, I'd expect it to be disabled by default and documented in types.go that it shouldn't be used because it's WIP and we expect to make breaking changes to it. I wouldn't want it, for example, to be shipped enabled in an alpha release and then broken later.

**Comment by kow3ns** [comment]
@janetkuo @bgrant0607 #49135

**Comment by janetkuo** [comment]
This is broken by #48544 which renamed auto-generated `DeepCopy` to `DeepCopyObject` :'(

```
# k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta2
vendor/k8s.io/api/apps/v1beta2/register.go:47: cannot use Deployment literal (type *Deployment) as type runtime.Object in argument to scheme.AddKnownTypes:
        *Deployment does not implement runtime.Object (missing DeepCopyObject method)
```

**Comment by janetkuo** [comment]
I fixed those failures due to recent changes in API machinery.

@bgrant0607 PTAL, mainly I submitted these two new commits:
* Don't enable apps/v1beta2 by default 4785e399c5e427827f2ef4cc596d2c99e0481e19
* Mark apps/v1beta2 as WIP in types.go 4b6397a0aadaa5f2f086a62500a724ae096ffd5f

**Comment by janetkuo** [comment]
/test pull-kubernetes-e2e-gce-etcd3

Seems like a test infra problem: #49320
```
I0720 19:13:34.368] Your "OAuth 2.0 Service Account" credentials are invalid. Please run
I0720 19:13:34.368]   $ gcloud auth login
I0720 19:13:34.368] Duplicate type [0:0:4]
I0720 19:13:46.852] CommandException: 1 file/object could not be transferred.
I0720 19:13:48.606] FAILED

W0720 19:13:48.738] 2017/07/20 19:13:48 main.go:220: Something went wrong: failed to acquire k8s binaries: error during /go/src/k8s.io/release/push-build.sh --nomock --verbose --noupdatelatest --bucket=kubernetes-release-pull --ci --gcs-suffix=/pull-kubernetes-e2e-gce-etcd3: exit status 2
```

**Comment by janetkuo** [comment]
Finally, all green again :smile:

File: pkg/master/master.go [hunk 2]
```diff
@@ -390,7 +389,8 @@ func DefaultAPIResourceConfigSource() *serverstorage.ResourceConfig {
 		authenticationv1beta1.SchemeGroupVersion,
 		autoscalingapiv1.SchemeGroupVersion,
 		appsv1beta1.SchemeGroupVersion,
-		appsv1beta2.SchemeGroupVersion,
+		// TODO: enable apps/v1beta2 by default before 1.8 release, after the API changes are done
+		// appsv1beta2.SchemeGroupVersion,
```
Comment (kow3ns) [inline pkg/master/master.go]: Turned off by default +1.

File: staging/src/k8s.io/api/apps/v1beta2/types.go [hunk 3]
```diff
@@ -29,13 +29,15 @@ const (
 	StatefulSetRevisionLabel       = ControllerRevisionHashLabelKey
 )
 
+// WIP: This is not ready to be used and we plan to make breaking changes to it.
```
Comment (kow3ns) [inline staging/src/k8s.io/api/apps/v1beta2/types.go]: Everything looks marked as WIP. +1

**Review by kow3ns — APPROVED** [review]
/lgtm

**Comment by bgrant0607** [comment]
Thanks much.
/approve

**Comment by janetkuo** [comment]
The failure comes from a test just merged: #49353

**Pull request merged**